kernel/aster-nix/src/fs/utils/dentrymnt.rs

```diff
@@ -0,0 +1,191 @@
+            ));
+        }
+        if self.mount_node.parent().is_some() & self.mount_node.mountpoint_dentry().is_some() {
+            let parent_dentrymnt = Self::new(
+                self.mount_node.parent().unwrap().upgrade().unwrap().clone(),
+                self.mount_node.mountpoint_dentry().unwrap().clone(),
+            );
+            parent_dentrymnt.effective_parent()
+        } else {
+            None
+        }
+    }
+
+    /// Get the overlaid dentrymnt of self.
+    ///
+    /// It will jump into the child mount if it is a mountpoint.
+    fn overlaid_dentrymnt(&self) -> Arc<Self> {
+        if !self.dentry.is_mountpoint() {
+            return self.this();
+        }
+        match self.mount_node.get(self) {
+            Some(child_mount) => Self::new(child_mount.clone(), child_mount.root_dentry().clone())
+                .overlaid_dentrymnt(),
+            None => self.this(),
+        }
+    }
+
+    /// Mount the fs on this dentrymnt. It will make this dentrymnt's dentry to be a mountpoint.
+    ///
+    /// If the given mountpoint has already been mounted, then its mounted child mount
+    /// will be updated.
+    /// The root dentry cannot be mounted.
+    ///
+    /// Return the mounted child mount.
+    pub fn mount(&self, fs: Arc<dyn FileSystem>) -> Result<Arc<MountNode>> {
+        if self.dentry.inode().type_() != InodeType::Dir {
+            return_errno!(Errno::ENOTDIR);
+        }
+        if self.effective_parent().is_none() {
+            return_errno_with_message!(Errno::EINVAL, "can not mount on root");
+        }
+        let child_mount = self.mount_node().mount(fs, &self.this())?;
+        self.dentry().set_mountpoint();
+        Ok(child_mount)
+    }
+
+    /// Unmount and return the mounted child mount.
+    ///
+    /// Note that the root mount cannot be unmounted.
+    pub fn umount(&self) -> Result<Arc<MountNode>> {
+        if !self.dentry.is_root_of_mount() {
+            return_errno_with_message!(Errno::EINVAL, "not mounted");
+        }
+
+        let mount_node = self.mount_node.clone();
+        let Some(mountpoint_dentry) = mount_node.mountpoint_dentry() else {
+            return_errno_with_message!(Errno::EINVAL, "cannot umount root mount");
+        };
+
+        let mountpoint_mount_node = mount_node.parent().unwrap().upgrade().unwrap();
+        let mountpoint_dentrymnt =
+            Self::new(mountpoint_mount_node.clone(), mountpoint_dentry.clone());
+
+        let child_mount = mountpoint_mount_node.umount(&mountpoint_dentrymnt)?;
+        mountpoint_dentry.clear_mountpoint();
+        Ok(child_mount)
+    }
+
+    /// Get the arc reference to self.
+    fn this(&self) -> Arc<Self> {
+        self.this.upgrade().unwrap()
+    }
+
+    /// Get the mount node of this dentrymnt.
+    pub fn mount_node(&self) -> &Arc<MountNode> {
+        &self.mount_node
+    }
+
+    /// Get the dentry of this dentrymnt.
+    pub fn dentry(&self) -> &Arc<Dentry> {
```

Comment:
I suggest hiding `Dentry` from the users of `DentryMnt`. Everything `Dentry` can do, it can be done through `DentryMnt`.

```rust
#[inherit_methods(from = "self.dentry")]
impl DentryMnt {
    pub fn fs(&self) -> Arc<dyn FileSystem>;
    pub fn sync(&self) -> Result<()>;
    pub fn metadata(&self) -> Metadata;
    pub fn type_(&self) -> InodeType;
    pub fn mode(&self) -> Result<InodeMode>;
    pub fn set_mode(&self, mode: InodeMode) -> Result<()>;
    pub fn size(&self) -> usize;
    pub fn resize(&self, size: usize) -> Result<()>;
    pub fn owner(&self) -> Result<Uid>;
    pub fn set_owner(&self, uid: Uid) -> Result<()>;
    pub fn group(&self) -> Result<Gid>;
    pub fn set_group(&self, gid: Gid) -> Result<()>;
    pub fn atime(&self) -> Duration;
    pub fn set_atime(&self, time: Duration);
    pub fn mtime(&self) -> Duration;
    pub fn set_mtime(&self, time: Duration);
}
```

In fact, I suggest (1) combing the two dentry modules into one, (2) exposing only a `Dentry` type (currently know as `DentryMnt`), which internally relies on a private `Dentry_` or `InnerDentry` (currently know as `Dentry`).

Reply:
Thank you very much for your guidance and suggestions. During the coding process, I noticed this issue. To implement certain methods of `DentryMnt`, I had to make methods defined in `Dentry` public, and these public methods have only one unique user, `DentryMnt`. 

Therefore, I have revisited the relationship between `DentryMnt` and `Dentry` and have merged these two modules. All the original methods in `Dentry` have been made private, with the necessary methods exposed (public) by `DentryMnt`, including methods such as `mknod`, `link`, `unlink`, and others, as well as methods inherited from `inode` by `dentry`. Regarding the naming of `DentryMnt` and `Dentry`, I believe that `DentryMnt` more clearly indicates that this structure is composed of `Dentry` and `MountNode`, as its main function is to uniquely represent a position in a directory tree using a tuple of a `Dentry` and a `MountNode`, rather than maintaining the hierarchy of the directory tree. Meanwhile, `Dentry` is primarily responsible for maintaining the hierarchy, so the name `Dentry` more clearly and intuitively represents a directory entry. 

I have also provided further explanations regarding the relationship between `DentryMnt`, `Dentry`, and `MountNode` in the latest comments.